Core/Core/Keychain/GeneralPurposeKeychain.swift

```diff
@@ -0,0 +1,119 @@
+//
+// This file is part of Canvas.
+// Copyright (C) 2019-present  Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import Foundation
+
+class GeneralPurposeKeychain {
+
+    private (set) public var serviceName: String
+    private (set) public var accessGroup: String?
+    static let shared = GeneralPurposeKeychain()
+
+    private static let defaultServiceName: String = {
+        return Bundle.main.bundleIdentifier ?? "com.instructure.general-purpose-keychain"
+    }()
+
+    private convenience init() {
+        self.init(serviceName: type(of: self).defaultServiceName)
+    }
+
+    init(serviceName: String, accessGroup: String? = nil) {
+        self.serviceName = serviceName
+        self.accessGroup = accessGroup
+    }
+
+    @discardableResult
+    func setItem(_ value: Any, for key: String) -> Bool {
+        let data = NSKeyedArchiver.archivedData(withRootObject: value)
```

Comment:
Do we need this to be NSCoding based? Or could we make use `Encodable` instead of of `Any`?

Reply:
My thinking was that NSCoding seems to handle dictionaries easier.  Also, with `Encodble` i get that ```Protocol type 'Encodable' cannot conform to 'Encodable' because only concrete types can conform to protocols``` dance.

Reply:
Yeah, you have to use a generic constraint for Encodable. `setItem<V: Encodable>(_ value: V,`

My concern is can't archivedData fail, or can it really archive `Any`? It doesn't look like it throws, so does that mean it just crashes?

Reply:
I've added a try catch